File: daily/2020-01-03/819 Most Common Word.py
class Solution(object):
    def mostCommonWord(self, paragraph, banned):
        import collections
        if paragraph == '':
            return
        for c in "!?',.:":
            paragraph = paragraph.replace(c, ' ')
        common = collections.Counter(map(lambda x: x.lower(), paragraph.split())).most_common()

        for i in common:
            key, fre = i
            if str(key) in banned:
                continue
            else:
                return key

# time complexity: o(P+B)
# space complexity: O(P+B)


    def mostCommonWord1(self, paragraph, banned):
        import re
        import collections
        ban = set(banned)
        words = re.findall(r'\w+', paragraph.lower())
        return collections.Counter(w for w in words if w not in ban).most_common(1)[0][0]
    

    def mostCommonWorld2(self, paragraph, banned):
        import collections
        if paragraph == '':
            return
        for c in '!?,.:':
            paragraph = paragraph.replace(c, ' ')
        common = collections.Counter(i for i in paragraph.split() if i not in banned).most_common(1)
        return common[0][0]

File: JavaScript Interview Code/819. Most Common Word.js
var mostCommonWord = function(paragraph, banned) {
    let dict = {}
    paragraph = paragraph.toLowerCase().split(/\W+/);

    let ans = '';
    let count = 0;
    for (let i=0; i<paragraph.length; i++) {
        key = paragraph[i];
        if (dict[key] == null) dict[key] = 0;
        dict[key] += 1;
    }

    for (key in dict) {
        if((dict[key] > count) && !(banned.includes(key))) {
            ans = key;
            count = dict[key];
        }
    }
    return ans;
}

// time complexity o(n)
// space complexity o(n)
